File: src/crawlers/lib/specific_crawlers/practice.py
import sys
from lib.tojson import exportJson
from lib.common import parse_url_content, get_driver_info

if __name__ == "__main__":
  year = sys.argv[1]
  location = sys.argv[2]
  id = sys.argv[3]
  practiceNb = sys.argv[4]

  # Formula 1 webpage for practice results
  url = f"https://www.formula1.com/en/results.html/{year}/races/{id}/{location}/practice-{practiceNb}.html"

  session_name, circuit_name, table = parse_url_content(url)

  results = []
  for row in table.find('tbody').find_all('tr'):
      columns = row.find_all('td')
      result = {
          "position": None if columns[1].text.strip() == "NC" else int(columns[1].text.strip()),
          "driver": get_driver_info(columns[3], columns[2]),
          "car": columns[4].text.strip(),
          "time": columns[5].text.strip(),
          "gap": columns[6].text.strip() if columns[6].text.strip() else None,
          "laps": int(columns[7].text.strip())
      }
      results.append(result)
  
  exportJson(
    data=results,
    id=id,
    location=location,
    circuit=circuit_name,
    session_name=session_name,
    data_type=f"fp{practiceNb}",
    exportname=f"fp{practiceNb}_results", 
    year=year
  )
  

File: src/jscrawlers/scripts/practice.ts
import * as cheerio from "cheerio";
import { extractDriverInfo, parseUrlContent } from "../lib/common";
import { writeData } from "../lib/export";
import { URL_BASE } from "../lib/constants";

interface PracticeResult {
  position: number;
  driver: Driver;
  car: string;
  time: string;
  gap: string;
  laps: number;
}

interface FetchPracticeParams {
  year: number | string;
  id: string;
  location: string;
  practiceNb: 1 | 2 | 3;
}

export async function fetchPracticeResults({
  year,
  id,
  location,
  practiceNb,
}: FetchPracticeParams) {
  try {
    const url = `${URL_BASE}/${year}/races/${id}/${location}/practice-${practiceNb}.html`;

    const { $, sessionName, circuit, tableRows } = await parseUrlContent(url);

    const results: PracticeResult[] = [];

    tableRows.each((_, row) => {
      const position = parseInt(
        $(row).find("td:nth-child(2)").text().trim(),
        10
      );
      const number = parseInt(
        $(row).find("td:nth-child(3)").text().trim(),
        10
      );
      const rawDriver = $(row)
        .find("td:nth-child(4)")
        .text()
        .trim()
        .replace(/\s+/g, " ");
      const driver = { ...extractDriverInfo(rawDriver), number };
      const car = $(row).find("td:nth-child(5)").text().trim();
      const time = $(row).find("td:nth-child(6)").text().trim();
      const gap = $(row).find("td:nth-child(7)").text().trim();
      const laps = parseInt(
        $(row).find("td:nth-child(8)").text().trim(),
        10
      );

      results.push({ position, driver, car, time, gap, laps });
    });

    writeData(results, {
      id,
      dataType: `fp${practiceNb}`,
      year,
      sessionName,
      circuit,
      location,
      fileName: `fp${practiceNb}_results.json`,
    });
    return results;
  } catch (error) {
    console.error("Error fetching the data:", error);
  }
}
